### PR TITLE
TF-610: Module filter email rules

### PR DESCRIPTION
**Pull request by ManhNTX**

### CODE REVIEW

File: rule_filter/.gitignore [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+# is commented out by default.
+#.vscode/
+
+# Flutter/Dart/Pub related
+# Libraries should not include pubspec.lock, per https://dart.dev/guides/libraries/private-files#pubspeclock.
+/pubspec.lock
+**/doc/api/
+.dart_tool/
+.packages
+build/
```
Comment (hoangdat) [inline rule_filter/.gitignore:30]: please add `*.mocks` also

File: rule_filter/lib/rule_filter/set/set_rule_filter_method.dart [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/capability/capability_identifier.dart';
+import 'package:jmap_dart_client/jmap/core/id.dart';
+import 'package:jmap_dart_client/jmap/core/method/request/set_method.dart';
+import 'package:jmap_dart_client/jmap/core/request/request_invocation.dart';
+import 'package:json_annotation/json_annotation.dart';
+import 'package:rule_filter/rule_filter/capability_rule_filter.dart';
+import 'package:rule_filter/rule_filter/tmail_rule.dart';
+
+class SetRuleFilterMethod extends SetMethod<List<TMailRule>> with OptionalUpdateRuleFilter<List<TMailRule>>{
```
Comment (hoangdat) [inline rule_filter/lib/rule_filter/set/set_rule_filter_method.dart:12]: out of scope

File: rule_filter/test/datasource/rule_filter_datasource_impl_test.mocks.dart [hunk 3]
```diff
@@ -0,0 +1,46 @@
+// Mocks generated by Mockito 5.2.0 from annotations
```
Comment (hoangdat) [inline rule_filter/test/datasource/rule_filter_datasource_impl_test.mocks.dart:1]: dont' push this file

File: rule_filter/lib/data/datasource_impl/rule_filter_datasource_impl.dart [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    });
+  }
+
+  @override
+  Future<void> updateListTMailRule(List<TMailRule> listTMailRule, AccountId accountId) {
+    return Future.sync(() async {
+       await _ruleFilterAPI.updateListTMailRule(listTMailRule, accountId);
+    }).catchError((error) {
+      throw error;
+    });  }
```
Comment (dab246) [inline rule_filter/lib/data/datasource_impl/rule_filter_datasource_impl.dart:28]: The closing quotation mark `}` of the function should get a newline

File: prebuild.sh [hunk 5]
```diff
@@ -14,6 +14,9 @@ flutter pub get && flutter pub run build_runner build --delete-conflicting-outpu
 cd ../contact
 flutter pub get && flutter pub run build_runner build --delete-conflicting-outputs
 
+cd ../rule_filter
```
Comment (dab246) [inline prebuild.sh:17]: Add to file `bitrise.yml`

File: env.file [hunk 6]
```diff
@@ -1,3 +1,3 @@
-SERVER_URL=http://localhost
+SERVER_URL=https://gateway.upn.integration-open-paas.org
```
Comment (hoangdat) [inline env.file:1]: please remove it

File: rule_filter/lib/data/datasource/rule_filter_datasource.dart [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:rule_filter/rule_filter/tmail_rule.dart';
+
+abstract class RuleFilterDataSource {
+  Future<List<TMailRule>> getListTMailRule(AccountId accountId);
+
+  Future<void> updateListTMailRule(List<TMailRule> listTMailRule, AccountId accountId);
```
Comment (hoangdat) [inline rule_filter/lib/data/datasource/rule_filter_datasource.dart:7]: I think you should separate `keepStateRules` and `updateRule` in parameters

File: rule_filter/lib/data/datasource_impl/rule_filter_datasource_impl.dart [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    List<TMailRule> listTMailRuleKeep,
+    List<TMailRule> listTMailRuleUpdate,
+    AccountId accountId,
+  ) {
+    final List<TMailRule> listTMailRule = List.empty(growable: true);
+    listTMailRule.addAll(listTMailRuleUpdate);
+    for (var itemTMailRuleKeep in listTMailRuleKeep) {
+      if (listTMailRuleUpdate
+          .where((itemTMailRuleUpdate) =>
+              itemTMailRuleKeep.id == itemTMailRuleUpdate.id)
```
Comment (hoangdat) [inline rule_filter/lib/data/datasource_impl/rule_filter_datasource_impl.dart:32]: I checked, but all the time I did not see the rule id return when I get it. Please compare the object to determine it is duplicated or not.

**Review by hoangdat — APPROVED** [review]

**Review by dab246 — APPROVED** [review]
/lgtm

**Pull request merged**